### ISSUE TITLE
Database schema changes

### ISSUE THREAD

**Issue by canweriotnow** [exercism/eio-api#4]
I'm going to start a checklist here to catch things I notice in the DB, and add to it as they come up. Check them off as we discuss/address them, not necessarily final resolution.

- [ ] case-insensitive (smash case) uniqueness constraint on username to prevent dupes (currently 31)
- [ ] foreign key constraints (currently none)
- [ ] use postgres JSON datatype when storing JSON blobs

**Comment by canweriotnow** [comment]
Just a note - if ActiveRecord migrations continue to be used, consider https://github.com/zoocasa/activerecord-postgresql-extensions

**Comment by kytrinyx** [comment]
_rephrasing from chat_

There's a good chance that we can simplify and get rid of a bunch of left-over concepts.

Back when I started out I didn't have a concept of an _exercise_ (e.g. a particular problem in a given track solved by a single user. An exercise has many submissions). I just had submissions. A lot of weird things are left-over from that.

The idea of "mute" got added in the first few weeks when there was no pagination. I was spending an inordinate amount of time nitpicking, and I kept clicking into submissions that I had already commented on. It was simply a way of hiding things that I had already seen.

This was back when I was messing with MongoDB, and I think that the separate table vs flag on the submissions is just an artifact of the migration (combined with not really having thought things through.

A better solution would be to have a list of all the exercises in reverse chronological order based on the newest activity (comment, submission). It should be paginated. It should have a read/unread marker for whether or not it has had new activity since you last looked at it. Like an email inbox, I suppose.

Another thing that was added in desperation rather than actually thinking through usability and interactions, was the idea of submission state (and later: exercise state). We have pending, hibernating, done, needs input, and maybe others. All of this was an attempt at not having an overwhelming number of exercises in the list waiting to nitpicked, combined with a kind of helpless/hopeless attempt at trying to make sure that people who haven't received any feedback will be prioritised.

This failed completely, to the point where there is now no real way to know if you've seen everything that has come in recently (if you nitpick a lot). Things get taken away from the main list in a way that is completely opaque to the user. This has some further negative effects in that it's hard to see discussions that are happening unless you accidentally get involved--meaning that it feels like you're in an empty room even if there's a lot going on. It also means that people who are very active might not see each-other's comments, and therefore track mentors might not notice that someone is giving really great feedback... which means that there's no pathway to bringing more people on as track mentors.

**Comment by canweriotnow** [comment]
@kytrinyx that's all really helpful... I think there's some room to condense tables and improve referential integrity, and maybe add some views to simplify queries that are purely UI-related, window functions to scope results sanely. Like the inbox-for-exercises concept.

Thinking deeply through submission states would be a good thing to do at some point, whether it should be an enum of mutually exclusive states, or possibly something better represented by a bitmask, e.g., something could be hibernating _and_ need input.

**Comment by kytrinyx** [comment]
Yeah, I've been wondering whether or not we could get rid of submission states altogether, but a bitmask might be more interesting.
